### PR TITLE
Add support for SM4 algorithm in CE instruction

### PR DESCRIPTION
**Pull request by tq444**
Qi Tao (1):
  cipher: add support for SM4(ECB) algorithm in CE instruction

Wenkai Lin (1):
  cipher: add support for SM4 CBC and CTR modes in CE instruction

Yang Shen (1):
  uadk/cipher: isa_ce - support SM4 cbc_cts mode

Yuzeng Zhuang (1):
  cipher: add support for SM4 CFB and XTS modes in CE instruction

 Makefile.am            |    5 +-
 drv/isa_ce_sm4.c       |  472 +++++++++
 drv/isa_ce_sm4.h       |   62 ++
 drv/isa_ce_sm4_armv8.S | 2296 ++++++++++++++++++++++++++++++++++++++++
 wd_cipher.c            |    4 +-
 5 files changed, 2835 insertions(+), 4 deletions(-)
 create mode 100644 drv/isa_ce_sm4.c
 create mode 100644 drv/isa_ce_sm4.h
 create mode 100644 drv/isa_ce_sm4_armv8.S

### CODE REVIEW

**Review by Liulongfang — COMMENTED** [review]
looks good to me

**Comment by lin755** [comment]
+1

File: drv/isa_ce_sm4.c [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+/*
+ * Copyright 2011-2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+/*
+ * Copyright 2023 Huawei Technologies Co.,Ltd. All rights reserved.
```
[the remark this answers is not in the document]
Comment (tq444) [reply drv/isa_ce_sm4.c]: 已修改

File: drv/isa_ce_sm4.c [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+ */
+
+#include "drv/wd_cipher_drv.h"
+#include "isa_ce_sm4.h"
+#include "wd_cipher.h"
```
[the remark this answers is not in the document]
Comment (tq444) [reply drv/isa_ce_sm4.c:15]: > 调整下顺序，从库到自己的顺序排列

已修改

File: drv/isa_ce_sm4.c [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+static void isa_ce_exit(struct wd_alg_driver *drv)
+{
+}
+
+/* increment upper 96 bits of 128-bit counter by 1 */
+static void ctr96_inc(__u8 *counter)
+{
+	__u32 n = INCREASE_BITS / BYTE_BITS;
```
[the remark this answers is not in the document]
Comment (tq444) [reply drv/isa_ce_sm4.c]: > 这里不是已经提了修改意见了吗？怎么还是相除。

已修改

**Review by Liulongfang — APPROVED** [review]
looks good to me

**Pull request merged**